### PR TITLE
ICU-22342 Implement ExternalBreakEngineAPI

### PR DESCRIPTION
**Pull request by FrankYFTang**
Implement the ExternalBreakEngineAPI and change the RBBI to use it internally. Add test cases to show how to do that in two fake break engines.

##### Checklist

- [X] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22342
- [X] Required: The PR title must be prefixed with a JIRA Issue number. 
- [X] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [X] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [x] Issue accepted (done by Technical Committee after discussion)
- [x] Tests included, if applicable
- [x] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

**Comment by FrankYFTang** [comment]
PING @aheninger @pedberg-icu

**Comment by FrankYFTang** [comment]
PING @aheninger @pedberg-icu @markusicu

**Comment by FrankYFTang** [comment]
ping

**Comment by FrankYFTang** [comment]
@richgillam It seems Peter is busy. I wonder would you have time to review this ?

File: icu4c/source/common/brkeng.cpp
Comment (richgillam) [inline icu4c/source/common/brkeng.cpp:350]: What's the unnamed `int32_t` parameter here?  From context, it looks like it'd be `startPos`.  If the caller passes you a `startPos`, why are you not using it?  Or is it something else?

File: icu4c/source/common/brkeng.cpp
Comment (richgillam) [inline icu4c/source/common/brkeng.cpp:140]: Probably want a blank line here...

**Review by richgillam — APPROVED** [review]
I don't know that I have the background to review this properly, but it all looks pretty straightforward and sensible to me...

**Review by pedberg-icu — APPROVED** [review]
Sorry for the late review. Looks reasonable.

**Review by aheninger — CHANGES_REQUESTED** [review]

File: icu4c/source/common/unicode/rbbi.h
Comment (FrankYFTang) [reply icu4c/source/common/unicode/rbbi.h:823]: so are you suggesting we do not provide a unregisterExternalBreakEngine method but free the object internally in u_cleanup()?

File: icu4c/source/common/brkeng.cpp
Comment (FrankYFTang) [reply icu4c/source/common/brkeng.cpp]: done

File: icu4c/source/common/unicode/rbbi.h
Comment (FrankYFTang) [reply icu4c/source/common/unicode/rbbi.h]: done

File: icu4c/source/common/brkeng.cpp
Comment (aheninger) [inline icu4c/source/common/brkeng.cpp:311]: This adopts `engine`, so it should probably delete it in case of failure. Use LocalPointers.

File: icu4c/source/common/brkeng.cpp
Comment (aheninger) [inline icu4c/source/common/brkeng.cpp:319]: Missing error checking for failed `new`

File: icu4c/source/common/brkeng.cpp
Comment (aheninger) [inline icu4c/source/common/brkeng.cpp:342]: Adopting constructors are troublesome for error handling. Ideally you would want to delete  `engine` on failure, but if the memory allocation for the new object fails, the constructor code won't be run, so you can't do it here.

LocalPointers can make it work, although it's a bit obscure. Look up passing std::unique_ptr as function parameters - they're very similar to LocalPointers.
Or, just deal with it at the call site.

File: icu4c/source/common/brkeng.cpp
Comment (aheninger) [inline icu4c/source/common/brkeng.cpp:138]: You could completely remove the `_deleteEngine()` function by making `LanguageBreakEngine` derive from `UObject` and using `uprv_deleteUObject()`

**Review by aheninger — CHANGES_REQUESTED** [review]
More to come, but a bit later.

File: icu4c/source/common/brkeng.cpp
Comment (aheninger) [inline icu4c/source/common/brkeng.cpp:386]: `foundBreaks.getBuffer()`, should be `foundBreaks.getBuffer()+beforeSize` ?
Looks like the code is setting up  to append, then writes at the start of `foundBreaks`.

File: icu4c/source/common/brkeng.cpp
Comment (FrankYFTang) [reply icu4c/source/common/brkeng.cpp:319]: Done by using LocalPointer now.

File: icu4c/source/test/intltest/rbbitst.cpp
Comment (FrankYFTang) [reply icu4c/source/test/intltest/rbbitst.cpp:5735]: fixe

File: icu4c/source/common/unicode/rbbi.h
Comment (srl295) [reply icu4c/source/common/unicode/rbbi.h:824]: agreed that u_cleanup() is the most important. Very unlikely for an app to make calls itself during shutdown. 

Unregister makes sense IF ICU can verify that it's not pulling the rug out from under users who already have created break instances.

File: icu4c/source/common/unicode/rbbi.h
Comment (srl295) [reply icu4c/source/common/unicode/rbbi.h:809]: Agreed. ICU should be able to invalidate its cache if need be here, on registry.

**Comment by srl295** [comment]
definitely support this idea and want to give this a more thorough review

**Comment by FrankYFTang** [comment]
See https://docs.google.com/document/d/1o6FDnf1fVZTETOPlQvwR_XrE_O1F--qIjGyhkCCVnI8/edit#heading=h.v4537jydx0ge

File: icu4c/source/common/unicode/rbbi.h
Comment (FrankYFTang) [reply icu4c/source/common/unicode/rbbi.h]: After ICUTC approve my amendment this morning I change this PR to remove the unregister function. PTAL

File: icu4c/source/common/unicode/rbbi.h
Comment (FrankYFTang) [reply icu4c/source/common/unicode/rbbi.h]: After ICUTC approve my amendment this morning I change this PR to mark them as "Technology Preview". PTAL

File: icu4c/source/common/brkeng.cpp
Comment (richgillam) [inline icu4c/source/common/brkeng.cpp:151]: Aren't you still going to need a mutex around this code?  It seems like if `loadEngineFor()` can execute while you're trying to read the array, you could have it change out from under you.  For that matter, it doesn't looks like `loadEngineFor()` takes the mutex either, which compounds the problem.  Or am I missing something?

File: icu4c/source/common/brkeng.h
Comment (richgillam) [inline icu4c/source/common/brkeng.h:92]: Should the second `ExternalLanguageBreakEngine` in this comment be `ExternalBreakEngine`?  That's what the code below seems to be saying.

Also, the naming here is kind of confusing: You have `LanguageBreakEngine`, `ExternalBreakEngine`, and `ExternalLanguageBreakEngine`, and it's not clear what their relationship is to one another.

File: icu4c/source/common/brkeng.cpp
Comment (richgillam) [inline icu4c/source/common/brkeng.cpp:308]: Don't you need a mutex around `fEngines` here (or around this whole function)?

**Review by richgillam — COMMENTED** [review]
Generally looks okay, but I have a few questions.

Also, I thought Andy had a block on this PR, but it looks like he removed it.  I'm still kind of hoping he can review this again before you check in.

File: icu4c/source/common/rbbi.cpp
Comment (FrankYFTang) [reply icu4c/source/common/rbbi.cpp:1269]: I change it to a common function.

**Review by richgillam — DISMISSED** [review]
Sold!  Thank you for the changes!

**Comment by FrankYFTang** [comment]
@aheninger @richgillam PTAL

**Review by richgillam — APPROVED** [review]
Rubber stamp, assuming nothing significant has changed since the last time I approved (can't really tell because the commits have been squashed).

**Pull request merged**
